### PR TITLE
windows support

### PR DESCRIPTION
**Pull request by mattn**

### CODE REVIEW

**Comment by justmao945** [comment]
Haha, this is a very nice patch. :+1:

File: plugin/clang.vim [hunk 1]
```diff
@@ -904,6 +914,8 @@ func! s:ClangExecute(root, clang_options, line, col)
   " longer version, redirect output to different files
   let l:command = l:cmd.' 1>'.l:tmps[0].' 2>'.l:tmps[1]
   let l:res = [[], []]
+  let b:clang_state['stdout'] = l:res[0]
+  let b:clang_state['stderr'] = l:res[1]
```
Comment (justmao945) [inline plugin/clang.vim]: I don't know why move them to here. They should be assigned when in sync mode, and useless in async mode in this function. So I think both of them should be kept where they were.

File: plugin/clang.vim [hunk 2]
```diff
@@ -941,16 +953,25 @@ func! s:ClangExecute(root, clang_options, line, col)
...
-    let l:keys = printf('ClangExecuteDone("%s","%s")', l:tmps[0], l:tmps[1])
-    let l:vcmd = printf('%s -s --noplugin --servername %s --remote-expr %s',
-                      \ g:clang_vim_exec, shellescape(v:servername), shellescape(l:keys))
-    let l:command = '('.l:command.';'.l:vcmd.') &'
-    call s:PDebug("s:ClangExecute::cmd", l:command, 2)
-    call system(l:command, l:src)
+    if s:is_windows
+      let l:keys = printf("ClangExecuteDone('%s','%s')", l:tmps[0], l:tmps[1])
+      let l:vcmd = printf('%s -s --noplugin --servername %s --remote-expr %s',
+                        \ g:clang_vim_exec, shellescape(v:servername), shellescape(l:keys))
```
Comment (justmao945) [inline plugin/clang.vim]: It seems `l:keys` and `l:vcmd` are similar in both under win and not. And I've tried the win version under OS X, it works well, so I recommend to use this version both in win and not win env. We should avoid duplicated code.
```viml
let l:keys = printf("ClangExecuteDone('%s','%s')", l:tmps[0], l:tmps[1])
let l:vcmd = printf('%s -s --noplugin --servername %s --remote-expr %s',
                        \ g:clang_vim_exec, shellescape(v:servername), shellescape(l:keys))
```

File: plugin/clang.vim [hunk 3]
```diff
@@ -941,16 +953,25 @@ func! s:ClangExecute(root, clang_options, line, col)
...
+      let l:input = tempname()
+      call writefile(split(l:src, "\n"), l:input)
```
Comment (justmao945) [inline plugin/clang.vim]: I think should keep empty line when split source.
```viml
call writefile(split(l:src, "\n", 1), l:input)
```

I don't know how it is working in your screenshot, but can't get the right completion when I use this case
```c++


struct Test {
int x;
int y;
};

int main() {
  Test t;
  t.
}

```

This case have leading empty lines, which will be removed when use split without keeping empty.

File: plugin/clang.vim [hunk 4]
```diff
@@ -941,16 +953,25 @@ func! s:ClangExecute(root, clang_options, line, col)
...
+      let l:command = 'type '.l:input.' | '.l:command.' && '.l:vcmd
```
Comment (justmao945) [inline plugin/clang.vim]: I think we should use `&` to combine clang and vim command. `&&` means vim command only be called when clang is successful, but `&` always calls the vim command , which is used to call the `ClangExecuteDone`. So we can display error diagnostics when have error.
```viml
let l:command = 'type '.l:input.' | '.l:command.' & '.l:vcmd
```

**Comment by justmao945** [comment]
Here's my comments, please tell me if I'm not right. :smile:

File: plugin/clang.vim (hunk 1 above)
Comment (mattn) [inline plugin/clang.vim]: Ah, okay. I'll fix.

File: plugin/clang.vim (hunk 2 above)
Comment (mattn) [inline plugin/clang.vim]: not same. on windows, cmd.exe has problem with double quoted strings. so it replaced double into single quote.

File: plugin/clang.vim (hunk 3 above)
Comment (mattn) [inline plugin/clang.vim]: okay, will fix.

File: plugin/clang.vim (hunk 4 above)
Comment (mattn) [inline plugin/clang.vim]: on windows, `;` doesn't work. because the first command eat it. it should be `&&`.

Comment (justmao945) [inline plugin/clang.vim]: Yes, `;` can't work, but `&` is what to replace it in cmd.exe.

bash | cmd.exe
--------|-----------
`;` | `&`
`&&` | `&&`

http://stackoverflow.com/questions/8055371/how-to-run-two-commands-in-one-line-in-windows-cmd

File: plugin/clang.vim (hunk 2 above)
Comment (justmao945) [inline plugin/clang.vim]: Yes, I know. I mean to use single quote version under all platforms, win, linux and OS X.
I've tested on OS X, it works well.

**Comment by mattn** [comment]
done. thank you for your review.

**Comment by justmao945** [comment]
Thank you for your work to support windows :smile:

**Pull request merged**